packages/webapp/webapp_server.js

```diff
@@ -571,92 +567,114 @@ function runWebAppServer() {
     syncQueue.runTask(function() {
       staticFilesByArch = Object.create(null);
 
-      function generateClientProgram(clientPath, arch) {
-        function addStaticFile(path, item) {
-          if (! hasOwn.call(staticFilesByArch, arch)) {
-            staticFilesByArch[arch] = Object.create(null);
-          }
-          staticFilesByArch[arch][path] = item;
-        }
+      const { configJson } = __meteor_bootstrap__;
+      const clientArchs = configJson.clientArchs ||
+        Object.keys(configJson.clientPaths);
 
-        // read the control for the client we'll be serving up
-        var clientJsonPath = pathJoin(__meteor_bootstrap__.serverDir,
-                                   clientPath);
-        var clientDir = pathDirname(clientJsonPath);
-        var clientJson = JSON.parse(readUtf8FileSync(clientJsonPath));
-        if (clientJson.format !== "web-program-pre1")
-          throw new Error("Unsupported format for client assets: " +
-                          JSON.stringify(clientJson.format));
-
-        if (! clientJsonPath || ! clientDir || ! clientJson)
-          throw new Error("Client config file not parsed.");
-
-        var manifest = clientJson.manifest;
-        _.each(manifest, function (item) {
-          if (item.url && item.where === "client") {
-            addStaticFile(getItemPathname(item.url), {
-              absolutePath: pathJoin(clientDir, item.path),
-              cacheable: item.cacheable,
-              hash: item.hash,
-              // Link from source to its map
-              sourceMapUrl: item.sourceMapUrl,
-              type: item.type
-            });
-
-            if (item.sourceMap) {
-              // Serve the source map too, under the specified URL. We assume all
-              // source maps are cacheable.
-              addStaticFile(getItemPathname(item.sourceMapUrl), {
-                absolutePath: pathJoin(clientDir, item.sourceMap),
-                cacheable: true
-              });
-            }
-          }
+      try {
+        clientArchs.forEach(arch => {
+          WebAppInternals.generateClientProgram(arch);
         });
+        WebAppInternals.staticFilesByArch = staticFilesByArch;
+      } catch (e) {
+        Log.error("Error reloading the client program: " + e.stack);
+        process.exit(1);
+      }
+    });
+  };
 
-        var program = {
-          format: "web-program-pre1",
-          manifest: manifest,
-          version: process.env.AUTOUPDATE_VERSION ||
-            WebAppHashing.calculateClientHash(
-              manifest,
-              null,
-              _.pick(__meteor_runtime_config__, "PUBLIC_SETTINGS")
-            ),
-          cordovaCompatibilityVersions: clientJson.cordovaCompatibilityVersions,
-          PUBLIC_SETTINGS: __meteor_runtime_config__.PUBLIC_SETTINGS
-        };
+  process.on("message", async (message) => {
+    if (message.package !== "webapp") return;
+    if (message.method === "generateClientProgram") {
```

Comment:
It's possible that the parent process might send this message before this handler has been registered, but that's ok, because the changes will get picked up when we call `WebAppInternals.reloadClientPrograms` for the first time (which happens later).